File: apps/mobile/lib/models/call.dart
enum CallType {
  voice,
  video,
}

enum CallStatus {
  incoming,
  outgoing,
  missed,
  declined,
  busy,
  failed,
}

class Call {
  final String id;
  final String callerId;
  final String callerName;
  final String? callerAvatar;
  final String receiverId;
  final String receiverName;
  final String? receiverAvatar;
  final CallType type;
  final CallStatus status;
  final DateTime timestamp;
  final int? duration; // 通话时长（秒）
  final bool isGroupCall;
  final List<String> participants;
  final String? roomId;
  final Map<String, dynamic>? metadata;

  Call({
    required this.id,
    required this.callerId,
    required this.callerName,
    this.callerAvatar,
    required this.receiverId,
    required this.receiverName,
    this.receiverAvatar,
    required this.type,
    required this.status,
    required this.timestamp,
    this.duration,
    this.isGroupCall = false,
    this.participants = const [],
    this.roomId,
    this.metadata,
  });

  Call copyWith({
    String? id,
    String? callerId,
    String? callerName,
    String? callerAvatar,
    String? receiverId,
    String? receiverName,
    String? receiverAvatar,
    CallType? type,
    CallStatus? status,
    DateTime? timestamp,
    int? duration,
    bool? isGroupCall,
    List<String>? participants,
    String? roomId,
    Map<String, dynamic>? metadata,
  }) {
    return Call(
      id: id ?? this.id,
      callerId: callerId ?? this.callerId,
      callerName: callerName ?? this.callerName,
      callerAvatar: callerAvatar ?? this.callerAvatar,
      receiverId: receiverId ?? this.receiverId,
      receiverName: receiverName ?? this.receiverName,
      receiverAvatar: receiverAvatar ?? this.receiverAvatar,
      type: type ?? this.type,
      status: status ?? this.status,
      timestamp: timestamp ?? this.timestamp,
      duration: duration ?? this.duration,
      isGroupCall: isGroupCall ?? this.isGroupCall,
      participants: participants ?? this.participants,
      roomId: roomId ?? this.roomId,
      metadata: metadata ?? this.metadata,
    );
  }

  Map<String, dynamic> toMap() {
    return {
      'id': id,
      'callerId': callerId,
      'callerName': callerName,
      'callerAvatar': callerAvatar,
      'receiverId': receiverId,
      'receiverName': receiverName,
      'receiverAvatar': receiverAvatar,
      'type': type.index,
      'status': status.index,
      'timestamp': timestamp.millisecondsSinceEpoch,
      'duration': duration,
      'isGroupCall': isGroupCall,
      'participants': participants,
      'roomId': roomId,
      'metadata': metadata,
    };
  }

  factory Call.fromMap(Map<String, dynamic> map) {
    return Call(
      id: map['id'] ?? '',
      callerId: map['callerId'] ?? '',
      callerName: map['callerName'] ?? '',
      callerAvatar: map['callerAvatar'],
      receiverId: map['receiverId'] ?? '',
      receiverName: map['receiverName'] ?? '',
      receiverAvatar: map['receiverAvatar'],
      type: CallType.values[map['type'] ?? 0],
      status: CallStatus.values[map['status'] ?? 0],
      timestamp: DateTime.fromMillisecondsSinceEpoch(map['timestamp']),
      duration: map['duration'],
      isGroupCall: map['isGroupCall'] ?? false,
      participants: List<String>.from(map['participants'] ?? []),
      roomId: map['roomId'],
      metadata: map['metadata'],
    );
  }

  bool get isVoiceCall => type == CallType.voice;
  bool get isVideoCall => type == CallType.video;
  bool get isIncoming => status == CallStatus.incoming;
  bool get isOutgoing => status == CallStatus.outgoing;
  bool get isMissed => status == CallStatus.missed;
  bool get isDeclined => status == CallStatus.declined;
  bool get isBusy => status == CallStatus.busy;
  bool get isFailed => status == CallStatus.failed;

  String get durationString {
    if (duration == null) return '';

    final minutes = duration! ~/ 60;
    final seconds = duration! % 60;

    if (minutes > 0) {
      return '$minutes:${seconds.toString().padLeft(2, '0')}';
    } else {
      return '${seconds}s';
    }
  }

  @override
  String toString() {
    return 'Call(id: $id, type: $type, status: $status, duration: $duration)';
  }

  @override
  bool operator ==(Object other) {
    if (identical(this, other)) return true;
    return other is Call && other.id == id;
  }

  @override
  int get hashCode => id.hashCode;
}

File: apps/mobile/src/domain/entities/Call.ts
import type { Call as DomainCall } from "@whatschat/shared-types";

export enum CallType {
  Voice = "voice",
  Video = "video",
}

export enum CallStatus {
  Incoming = "incoming",
  Outgoing = "outgoing",
  Missed = "missed",
  Declined = "declined",
  Busy = "busy",
  Failed = "failed",
}

export interface Call extends Omit<DomainCall, "type" | "status"> {
  type: CallType;
  status: CallStatus;
  callerId: string;
  callerName: string;
  receiverId: string;
  receiverName: string;
  timestamp: Date;
  isGroupCall: boolean;
  participants: string[];
}

export class CallEntity implements Call {
  id: string;
  callerId: string;
  callerName: string;
  callerAvatar?: string;
  receiverId: string;
  receiverName: string;
  receiverAvatar?: string;
  type: CallType;
  status: CallStatus;
  timestamp: Date;
  duration?: number;
  isGroupCall: boolean;
  participants: string[];
  roomId?: string;
  metadata?: Record<string, unknown>;

  constructor(data: Call) {
    this.id = data.id;
    this.callerId = data.callerId;
    this.callerName = data.callerName;
    this.callerAvatar = data.callerAvatar;
    this.receiverId = data.receiverId;
    this.receiverName = data.receiverName;
    this.receiverAvatar = data.receiverAvatar;
    this.type = data.type;
    this.status = data.status;
    this.timestamp = data.timestamp;
    this.duration = data.duration;
    this.isGroupCall = data.isGroupCall ?? false;
    this.participants = data.participants ?? [];
    this.roomId = data.roomId;
    this.metadata = data.metadata;
  }

  copyWith(updates: Partial<Call>): CallEntity {
    return new CallEntity({
      id: updates.id ?? this.id,
      callerId: updates.callerId ?? this.callerId,
      callerName: updates.callerName ?? this.callerName,
      callerAvatar: updates.callerAvatar ?? this.callerAvatar,
      receiverId: updates.receiverId ?? this.receiverId,
      receiverName: updates.receiverName ?? this.receiverName,
      receiverAvatar: updates.receiverAvatar ?? this.receiverAvatar,
      type: updates.type ?? this.type,
      status: updates.status ?? this.status,
      timestamp: updates.timestamp ?? this.timestamp,
      duration: updates.duration ?? this.duration,
      isGroupCall: updates.isGroupCall ?? this.isGroupCall,
      participants: updates.participants ?? this.participants,
      roomId: updates.roomId ?? this.roomId,
      metadata: updates.metadata ?? this.metadata,
    });
  }

  toMap(): Record<string, unknown> {
    return {
      id: this.id,
      callerId: this.callerId,
      callerName: this.callerName,
      callerAvatar: this.callerAvatar,
      receiverId: this.receiverId,
      receiverName: this.receiverName,
      receiverAvatar: this.receiverAvatar,
      type: this.type,
      status: this.status,
      timestamp: this.timestamp.getTime(),
      duration: this.duration,
      isGroupCall: this.isGroupCall,
      participants: this.participants,
      roomId: this.roomId,
      metadata: this.metadata,
    };
  }

  static fromMap(map: Record<string, unknown>): CallEntity {
    const callTypeMap: Record<number, CallType> = {
      0: CallType.Voice,
      1: CallType.Video,
    };

    const callStatusMap: Record<number, CallStatus> = {
      0: CallStatus.Incoming,
      1: CallStatus.Outgoing,
      2: CallStatus.Missed,
      3: CallStatus.Declined,
      4: CallStatus.Busy,
      5: CallStatus.Failed,
    };

    return new CallEntity({
      id: (map.id as string) ?? "",
      callerId: (map.callerId as string) ?? "",
      callerName: (map.callerName as string) ?? "",
      callerAvatar: map.callerAvatar as string | undefined,
      receiverId: (map.receiverId as string) ?? "",
      receiverName: (map.receiverName as string) ?? "",
      receiverAvatar: map.receiverAvatar as string | undefined,
      type: callTypeMap[(map.type as number) ?? 0] ?? CallType.Voice,
      status: callStatusMap[(map.status as number) ?? 0] ?? CallStatus.Incoming,
      timestamp: new Date(map.timestamp as number),
      duration: map.duration as number | undefined,
      isGroupCall: (map.isGroupCall as boolean) ?? false,
      participants: Array.isArray(map.participants) ? (map.participants as string[]) : [],
      roomId: map.roomId as string | undefined,
      metadata: map.metadata as Record<string, unknown> | undefined,
    });
  }

  get isVoiceCall(): boolean {
    return this.type === CallType.Voice;
  }

  get isVideoCall(): boolean {
    return this.type === CallType.Video;
  }

  get isIncoming(): boolean {
    return this.status === CallStatus.Incoming;
  }

  get isOutgoing(): boolean {
    return this.status === CallStatus.Outgoing;
  }

  get isMissed(): boolean {
    return this.status === CallStatus.Missed;
  }

  get isDeclined(): boolean {
    return this.status === CallStatus.Declined;
  }

  get isBusy(): boolean {
    return this.status === CallStatus.Busy;
  }

  get isFailed(): boolean {
    return this.status === CallStatus.Failed;
  }

  get durationString(): string {
    if (this.duration == null) return "";

    const minutes = Math.floor(this.duration / 60);
    const seconds = this.duration % 60;

    if (minutes > 0) {
      return `${minutes}:${seconds.toString().padStart(2, "0")}`;
    } else {
      return `${seconds}s`;
    }
  }

  toString(): string {
    return `Call(id: ${this.id}, type: ${this.type}, status: ${this.status}, duration: ${this.duration})`;
  }

  equals(other: Call): boolean {
    return this.id === other.id;
  }
}
